File: mobile/src/services/equipment/EquipmentService.ts
import DatabaseService from '../database/DatabaseService';
import {Equipment, EquipmentType} from '../../models/Equipment';
import {v4 as uuidv4} from 'uuid';

class EquipmentService {
  async getAllEquipment(): Promise<Equipment[]> {
    const db = DatabaseService.getDatabase();
    const [results] = await db.executeSql('SELECT * FROM equipment');
    return results.rows.raw().map(this.mapRowToEquipment);
  }

  async getEquipmentById(id: string): Promise<Equipment | null> {
    const db = DatabaseService.getDatabase();
    const [results] = await db.executeSql(
      'SELECT * FROM equipment WHERE id = ?',
      [id],
    );
    if (results.rows.length === 0) {
      return null;
    }
    return this.mapRowToEquipment(results.rows.item(0));
  }

  async createEquipment(equipment: Omit<Equipment, 'id'>): Promise<Equipment> {
    const db = DatabaseService.getDatabase();
    const id = uuidv4();

    await db.executeSql(
      `INSERT INTO equipment (id, name, type, manufacturer, model, serial_number, purchase_date, notes)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        id,
        equipment.name,
        equipment.type,
        equipment.manufacturer || null,
        equipment.model || null,
        equipment.serialNumber || null,
        equipment.purchaseDate?.toISOString() || null,
        equipment.notes || null,
      ],
    );

    return this.getEquipmentById(id) as Promise<Equipment>;
  }

  async updateEquipment(
    id: string,
    updates: Partial<Equipment>,
  ): Promise<Equipment> {
    const db = DatabaseService.getDatabase();
    // Implementation for update
    return this.getEquipmentById(id) as Promise<Equipment>;
  }

  async deleteEquipment(id: string): Promise<void> {
    const db = DatabaseService.getDatabase();
    await db.executeSql('DELETE FROM equipment WHERE id = ?', [id]);
  }

  private mapRowToEquipment(row: any): Equipment {
    return {
      id: row.id,
      name: row.name,
      type: row.type as EquipmentType,
      manufacturer: row.manufacturer || undefined,
      model: row.model || undefined,
      serialNumber: row.serial_number || undefined,
      purchaseDate: row.purchase_date ? new Date(row.purchase_date) : undefined,
      notes: row.notes || undefined,
    };
  }
}

export default new EquipmentService();

File: frontend/lib/services/equipment_service.dart
import 'package:uuid/uuid.dart';
import '../models/equipment.dart';
import 'database_service.dart';

class EquipmentService {
  final DatabaseService _db = DatabaseService();
  final _uuid = const Uuid();

  Future<String> createEquipment({
    required String name,
    required EquipmentType type,
    String? manufacturer,
    String? model,
    String? serialNumber,
    DateTime? purchaseDate,
    String? notes,
  }) async {
    final equipment = Equipment(
      id: _uuid.v4(),
      name: name,
      type: type,
      manufacturer: manufacturer,
      model: model,
      serialNumber: serialNumber,
      purchaseDate: purchaseDate,
      notes: notes,
    );

    await _db.insertEquipment(equipment);
    return equipment.id;
  }

  Future<List<Equipment>> getAllEquipment() async {
    return await _db.getAllEquipment();
  }

  Future<Equipment?> getEquipmentById(String id) async {
    return await _db.getEquipmentById(id);
  }

  Future<void> updateEquipment(Equipment equipment) async {
    await _db.updateEquipment(equipment);
  }

  Future<void> deleteEquipment(String id) async {
    await _db.deleteEquipment(id);
  }

  List<Equipment> filterByType(List<Equipment> equipment, EquipmentType type) {
    return equipment.where((e) => e.type == type).toList();
  }
}
